### PR TITLE
Implement SANSA model (scalable variant of EASE)

### PR DESCRIPTION
**Pull request by matospiso**
### Description

Implement a scalable (approximate) variant of EASE: **SANSA**
Paper: https://dl.acm.org/doi/10.1145/3604915.3608827

### Related Issues

https://github.com/PreferredAI/cornac/issues/654

### Checklist:

- [ ] I have added tests.
- [ ] I have updated the documentation accordingly.
- [x] I have updated `README.md` (if you are adding a new model).
- [x] I have updated `examples/README.md` (if you are adding a new example).
- [ ] I have updated `datasets/README.md` (if you are adding a new dataset).

### CODE REVIEW

**Comment by matospiso** [comment]
> I have updated the documentation accordingly.

Not sure about this point.

**Comment by tqtg** [comment]
> > I have updated the documentation accordingly.
> 
> Not sure about this point.

Ah, we should've removed this one. Previously, we need to add a model entry for documentation manually. We have now automated the parsing from README.md file.

**Comment by tqtg** [comment]
Thanks @matospiso for this PR. Do we have a quick comparison between this version and the original version of EASE?

File: cornac/models/sansa/recom_sansa.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
+import numpy as np
+import scipy.sparse as sp
+
+from sansa.core import (
```
Comment (tqtg) [inline cornac/models/sansa/recom_sansa.py:4]: We need to put all these import statements from sansa package under `fit` function. By default, we do not have sansa installed so it will cause import errors.

File: cornac/models/sansa/recom_sansa.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    FactorizationMethod,
+    GramianFactorizer,
+    CHOLMODGramianFactorizerConfig,
+    ICFGramianFactorizerConfig,
+    UnitLowerTriangleInverter,
+    UMRUnitLowerTriangleInverterConfig,
+)
+from sansa.utils import get_squared_norms_along_compressed_axis, inplace_scale_along_compressed_axis, inplace_scale_along_uncompressed_axis
```
Comment (tqtg) [inline cornac/models/sansa/recom_sansa.py:12]: same as the above

File: cornac/models/sansa/recom_sansa.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+        factorizer_shift_multiplier=2.0,
+        inverter_scans=3,
+        inverter_finetune_steps=10,
+        use_absolute_value_scores=False,
+        trainable=True,
+        verbose=True,
+        seed=None,
+        W1=None,  # "weights[0] (sp.csr_matrix)"
+        W2=None,  # "weights[1] (sp.csr_matrix)"
+        X=None,  # user-item interaction matrix (sp.csr_matrix)
```
Comment (tqtg) [inline cornac/models/sansa/recom_sansa.py:105]: We shouldn't have `X` here. Training data should have passed in `fit` function via `train_set`.

File: cornac/models/sansa/recom_sansa.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        self.compute_gramian = compute_gramian
+        self.use_absolute_value_scores = use_absolute_value_scores
+        self.verbose = verbose
+        self.seed = seed
+        self.X = X
+        if self.X is not None:
+            self.X = self.X.astype(np.float32)
+        self.weights = (W1, W2)
+
+        if factorizer_class == "CHOLMOD":
```
Comment (tqtg) [inline cornac/models/sansa/recom_sansa.py:119]: Initialization for `factorizer` and `inverter` can be inside `fit` function. We probably do not need all these objects after training. Only weights are necessary for inference?

File: cornac/models/sansa/recom_sansa.py (hunk 1 above)
Comment (matospiso) [reply cornac/models/sansa/recom_sansa.py]: changed

File: cornac/models/sansa/recom_sansa.py (hunk 2 above)
Comment (matospiso) [reply cornac/models/sansa/recom_sansa.py]: changed

File: cornac/models/sansa/recom_sansa.py (hunk 4 above)
Comment (matospiso) [reply cornac/models/sansa/recom_sansa.py]: changed

File: cornac/models/sansa/recom_sansa.py (hunk 3 above)
Comment (matospiso) [reply cornac/models/sansa/recom_sansa.py:95]: sure - but how does inference work for validation/test users under strong generalization in this framework?
I copied the `score` function from EASE, but it works with user_idx instead of interaction vectors as I would expect.
EASE (and other similar models) can infer scores from interaction vectors even for users not seen during training.

**Comment by matospiso** [comment]
> Do we have a quick comparison between this version and the original version of EASE?

Do you mean in some specific sense? Accuracy, or resource consumption?

File: cornac/models/sansa/recom_sansa.py (hunk 3 above)
Comment (tqtg) [reply cornac/models/sansa/recom_sansa.py:95]: I got that the model needs interactions to make prediction. In Cornac, it will go through an Experiment with train/val/test sets available. In this event, you've already stored the interactions `X` during the call of `fit`, and it will be reused in the `score` function. For `unknown`/`cold-start` users, we do need a mechanism to output scores, e.g., we check that as in [here](https://github.com/PreferredAI/cornac/blob/167a212f3144eb1016829fdcc76f5217f5b471cc/cornac/models/mf/recom_mf.py#L282). For the case of EASE, I think we output the default item values in `B` for unknown users?

**Comment by tqtg** [comment]
> > Do we have a quick comparison between this version and the original version of EASE?
> 
> Do you mean in some specific sense? Accuracy, or resource consumption?

I mean you've already created examples to run the model. Can we compare recommendation performance of EASE and SANSA on the MovieLens dataset? Just put the numbers here so we know that they're not very far off :)

File: cornac/models/sansa/recom_sansa.py (hunk 3 above)
Comment (matospiso) [reply cornac/models/sansa/recom_sansa.py:95]: I thought the validation/test evaluation was done in a way that users are disjoint between splits, and that's why I added the `X` - so that we can replace it after training with interactions of validation/test users (keeping the weights) and use these new rows as inference inputs. 

I assumed that's why EASE _also_ passes the user-item interaction matrix as an optional argument to `__init__`:
```
    def __init__(
        self,
        name="EASEᴿ",
        lamb=500,
        posB=True,
        trainable=True,
        verbose=True,
        seed=None,
        B=None,
        U=None,
    ):
        Recommender.__init__(self, name=name, trainable=trainable, verbose=verbose)
        self.lamb = lamb
        self.posB = posB
        self.verbose = verbose
        self.seed = seed
        self.B = B
        self.U = U  # this is the same as X in SANSA
```
(see [recom_ease.py](https://github.com/PreferredAI/cornac/blob/master/cornac/models/ease/recom_ease.py#L38-L55) ). 

So I pretty much just copied the interface of EASE in my implementation, but we can change it if you think it's better to remove the `X`.

Comment (tqtg) [reply cornac/models/sansa/recom_sansa.py:95]: I think the confusion is that we're using [RatioSplit](https://cornac.readthedocs.io/en/stable/api_ref/eval_methods.html#module-cornac.eval_methods.ratio_split) which is referred to as `weak generalization` in the EASE paper. The author evaluated the models based on `strong generalization` scheme which is similar to user-based [StratifiedSplit](https://cornac.readthedocs.io/en/stable/api_ref/eval_methods.html#module-cornac.eval_methods.stratified_split). For the purpose of introducing SANSA model into Cornac, this implementation is fine. However, if we really insist on replicating the author's implementation, we need to further work on the correct evaluation method. I'm happy to merge this PR for now.

**Review by tqtg — APPROVED** [review]

File: cornac/models/sansa/recom_sansa.py (hunk 3 above)
Comment (matospiso) [reply cornac/models/sansa/recom_sansa.py:95]: yes, that's exactly why I was confused - I thought the EASE implementation already supports StratifiedSplit using this "hack"

It should be pretty easy to add support for this split method to both EASE and SANSA in the future :)
